File: mt5-core-server/src/services/auto/analyzers/smc/sweepDetection.ts
/**
 * SMC Engine V2 — Multi-TF Sweep Detection + Attack Force
 * Port จาก Pine Script V8.3
 * 2026-05-06
 */

import type { Candle, MtfSweep, MtfSweepScore, AttackForce } from './types.js';
import { computeATR, bodySize, candleMax, candleMin, isBullish, isBearish } from './utils.js';

// ── Sweep Detection ────────────────────────────────────────────────

const RECLAIM_FACTOR = 0.50;
const MIN_WICK_ATR_MULT = 0.30;

/**
 * Detect sweeps on a single timeframe candle array.
 * Port of Pine Script f_sweepSignals()
 */
export function detectSweeps(candles: Candle[]): { bull: boolean; bear: boolean } {
  if (candles.length < 10) return { bull: false, bear: false };

  const last = candles[candles.length - 1];
  const atr = computeATR(candles);

  // Find nearest bearish candle (for bullish OB)
  let bullTop: number | null = null, bullBtm: number | null = null;
  let bearTop: number | null = null, bearBtm: number | null = null;

  for (let i = candles.length - 2; i >= 0; i--) {
    if (isBearish(candles[i]) && bullTop === null) {
      bullTop = candleMax(candles[i]);
      bullBtm = candleMin(candles[i]);
    }
    if (isBullish(candles[i]) && bearTop === null) {
      bearTop = candleMax(candles[i]);
      bearBtm = candleMin(candles[i]);
    }
    if (bullTop !== null && bearTop !== null) break;
  }

  let bull = false, bear = false;

  if (bullBtm !== null && bullTop !== null) {
    const width = Math.abs(bullTop - bullBtm);
    const reclaim = bullBtm + RECLAIM_FACTOR * width;
    const wick = bullBtm - last.l;
    bull = width > 0 && last.l < bullBtm && last.c > reclaim && wick > (MIN_WICK_ATR_MULT * atr);
  }

  if (bearTop !== null && bearBtm !== null) {
    const width = Math.abs(bearTop - bearBtm);
    const reclaim = bearTop - RECLAIM_FACTOR * width;
    const wick = last.h - bearTop;
    bear = width > 0 && last.h > bearTop && last.c < reclaim && wick > (MIN_WICK_ATR_MULT * atr);
  }

  return { bull, bear };
}

/**
 * Aggregate MTF sweeps from multiple timeframe candle arrays.
 * Similar to Pine Script's aggregate mode (M1+M5+M15 score >= 2)
 */
export function aggregateMtfSweeps(
  tfCandles: { timeframe: string; candles: Candle[] }[]
): MtfSweepScore {
  const sweeps: MtfSweep[] = [];
  let bullScore = 0, bearScore = 0;

  for (const { timeframe, candles } of tfCandles) {
    if (candles.length < 10) continue;
    const { bull, bear } = detectSweeps(candles);
    const last = candles[candles.length - 1];

    if (bull) {
      bullScore++;
      sweeps.push({ timeframe, side: 'BULL', barIndex: candles.length - 1, price: last.c });
    }
    if (bear) {
      bearScore++;
      sweeps.push({ timeframe, side: 'BEAR', barIndex: candles.length - 1, price: last.c });
    }
  }

  let aggregateSide: 'BULL' | 'BEAR' | 'NEUTRAL' = 'NEUTRAL';
  if (bullScore >= 2 && bullScore > bearScore) aggregateSide = 'BULL';
  else if (bearScore >= 2 && bearScore > bullScore) aggregateSide = 'BEAR';

  return { bullScore, bearScore, sweeps, aggregateSide };
}

// ── Attack Force Detection ─────────────────────────────────────────

export interface AttackForceConfig {
  atrMultiplier: number;     // Body must be > ATR * mult (default: 2.0)
  bodyAvgPeriod: number;     // SMA period for body size (default: 10)
  volumeAvgPeriod: number;   // SMA period for volume (default: 10)
  bodyMultiplier: number;    // Body must be > avgBody * mult (default: 2.0)
  volumeMultiplier: number;  // Volume must be > avgVol * mult (default: 1.5)
}

const DEFAULT_AF_CFG: AttackForceConfig = {
  atrMultiplier: 2.0,
  bodyAvgPeriod: 10,
  volumeAvgPeriod: 10,
  bodyMultiplier: 2.0,
  volumeMultiplier: 1.5,
};

/**
 * Detect high-momentum "Attack Force" candles.
 * Port of Pine Script's Attack Force logic.
 */
export function detectAttackForce(
  candles: Candle[],
  config: Partial<AttackForceConfig> = {}
): AttackForce[] {
  const cfg = { ...DEFAULT_AF_CFG, ...config };
  if (candles.length < cfg.bodyAvgPeriod + 1) return [];

  const atr = computeATR(candles);
  const forces: AttackForce[] = [];

  // Only check recent candles (last 5)
  const startIdx = Math.max(cfg.bodyAvgPeriod, candles.length - 5);

  for (let i = startIdx; i < candles.length; i++) {
    const c = candles[i];
    const body = bodySize(c);

    // Average body size (SMA)
    let bodySum = 0;
    for (let j = i - cfg.bodyAvgPeriod; j < i; j++) {
      bodySum += bodySize(candles[j]);
    }
    const avgBody = bodySum / cfg.bodyAvgPeriod;

    // Average volume (SMA)
    let volSum = 0;
    const volStart = Math.max(0, i - cfg.volumeAvgPeriod);
    const volCount = i - volStart;
    for (let j = volStart; j < i; j++) {
      volSum += candles[j].v;
    }
    const avgVol = volCount > 0 ? volSum / volCount : 0;

    const bodyCondition = body > avgBody * cfg.bodyMultiplier;
    const volCondition = avgVol <= 0 || c.v > avgVol * cfg.volumeMultiplier;
    const atrCondition = body > atr * cfg.atrMultiplier;

    if (bodyCondition && volCondition && atrCondition) {
      forces.push({
        barIndex: i,
        side: isBullish(c) ? 'BULL' : 'BEAR',
        bodySize: body,
        atrRatio: atr > 0 ? body / atr : 0,
        volumeRatio: avgVol > 0 ? c.v / avgVol : 0,
      });
    }
  }

  return forces;
}

File: composeApp/src/commonMain/kotlin/com/example/personalaibot/automation/smc/SweepDetection.kt
package com.example.personalaibot.automation.smc

import com.example.personalaibot.tools.trading.Candle
import kotlin.math.abs
import kotlin.math.max

/**
 * Sweep Detection + Attack Force — port จาก mt5-core-server sweepDetection.ts (Pine V8.3)
 * บนมือถือมี TF เดียวต่อ job — ใช้ detectSweeps แบบ single-TF (MTF aggregate ของต้นฉบับข้ามไป)
 */
internal object SweepDetection {

    private const val RECLAIM_FACTOR = 0.50
    private const val MIN_WICK_ATR_MULT = 0.30

    /** (bull, bear) — sweep ของแท่งล่าสุดเทียบกับแท่ง bearish/bullish ที่ใกล้สุด */
    fun detectSweeps(candles: List<Candle>): Pair<Boolean, Boolean> {
        if (candles.size < 10) return false to false

        val last = candles[candles.size - 1]
        val atr = computeAtr(candles)

        var bullTop: Double? = null
        var bullBtm: Double? = null
        var bearTop: Double? = null
        var bearBtm: Double? = null

        for (i in candles.size - 2 downTo 0) {
            if (isBearish(candles[i]) && bullTop == null) {
                bullTop = candleMax(candles[i]); bullBtm = candleMin(candles[i])
            }
            if (isBullish(candles[i]) && bearTop == null) {
                bearTop = candleMax(candles[i]); bearBtm = candleMin(candles[i])
            }
            if (bullTop != null && bearTop != null) break
        }

        var bull = false
        var bear = false

        val bb = bullBtm; val bt = bullTop
        if (bb != null && bt != null) {
            val width = abs(bt - bb)
            val reclaim = bb + RECLAIM_FACTOR * width
            val wick = bb - last.low
            bull = width > 0 && last.low < bb && last.close > reclaim && wick > MIN_WICK_ATR_MULT * atr
        }

        val rb = bearBtm; val rt = bearTop
        if (rb != null && rt != null) {
            val width = abs(rt - rb)
            val reclaim = rt - RECLAIM_FACTOR * width
            val wick = last.high - rt
            bear = width > 0 && last.high > rt && last.close < reclaim && wick > MIN_WICK_ATR_MULT * atr
        }

        return bull to bear
    }

    /** Candle-confirmed liquidity event: sweep = pierce + reclaim; run = decisive close beyond liquidity. */
    fun detectLiquidityEvents(candles: List<Candle>, zones: List<LiquidityZone>): Pair<SmcSignalConfirmation, SmcSignalConfirmation> {
        if (candles.size < 2 || zones.isEmpty()) return SmcSignalConfirmation.NONE to SmcSignalConfirmation.NONE
        val c = candles.last()
        val atr = computeAtr(candles)
        val tol = max(atr * 0.10, (c.high - c.low) * 0.05)
        val above = zones.filter { it.isHigh && it.price > 0 }.minByOrNull { abs(it.price - c.high) }
        val below = zones.filter { !it.isHigh && it.price > 0 }.minByOrNull { abs(it.price - c.low) }
        fun event(level: Double, highSide: Boolean): SmcSignalConfirmation {
            val pierced = if (highSide) c.high > level + tol else c.low < level - tol
            if (!pierced) return SmcSignalConfirmation.NONE
            val reclaimed = if (highSide) c.close < level else c.close > level
            return if (reclaimed) SmcSignalConfirmation.LIQUIDITY_SWEEP else SmcSignalConfirmation.LIQUIDITY_RUN
        }
        return (below?.let { event(it.price, false) } ?: SmcSignalConfirmation.NONE) to
            (above?.let { event(it.price, true) } ?: SmcSignalConfirmation.NONE)
    }

    /** Attack Force — แท่งโมเมนตัมแรงผิดปกติ (body > 2×ATR และ > 2×ค่าเฉลี่ย + volume ยืนยัน) 5 แท่งล่าสุด */
    fun detectAttackForce(
        candles: List<Candle>,
        atrMultiplier: Double = 2.0,
        bodyAvgPeriod: Int = 10,
        volumeAvgPeriod: Int = 10,
        bodyMultiplier: Double = 2.0,
        volumeMultiplier: Double = 1.5
    ): List<AttackForce> {
        if (candles.size < bodyAvgPeriod + 1) return emptyList()

        val atr = computeAtr(candles)
        val forces = mutableListOf<AttackForce>()
        val startIdx = max(bodyAvgPeriod, candles.size - 5)

        for (i in startIdx until candles.size) {
            val c = candles[i]
            val body = bodySize(c)

            var bodySum = 0.0
            for (j in i - bodyAvgPeriod until i) bodySum += bodySize(candles[j])
            val avgBody = bodySum / bodyAvgPeriod

            var volSum = 0.0
            val volStart = max(0, i - volumeAvgPeriod)
            for (j in volStart until i) volSum += candles[j].volume
            val volCount = i - volStart
            val avgVol = if (volCount > 0) volSum / volCount else 0.0

            val bodyOk = body > avgBody * bodyMultiplier
            val volOk = avgVol <= 0 || c.volume > avgVol * volumeMultiplier
            val atrOk = body > atr * atrMultiplier

            if (bodyOk && volOk && atrOk) {
                forces += AttackForce(
                    barIndex = i,
                    isBull = isBullish(c),
                    bodySize = body,
                    atrRatio = if (atr > 0) body / atr else 0.0,
                    volumeRatio = if (avgVol > 0) c.volume / avgVol else 0.0
                )
            }
        }

        return forces
    }
}
